internal/controllers/token_exchange.go

```diff
@@ -276,59 +279,58 @@ func (t *TokenExchangeController) evaluateSacdDoc(c *fiber.Ctx, record *models.P
 
 	// If any permissions are missing, return an error
 	if len(missingPermissions) > 0 {
-		return fiber.NewError(fiber.StatusBadRequest,
-			fmt.Sprintf("Address %s lacks permissions %v on token id %d for asset %s.",
-				grantee.Hex(), missingPermissions, pr.TokenID, pr.NFTContractAddress))
+		return fmt.Errorf("missing permissions: %v on token id %d for asset %s", missingPermissions, tokenReq.TokenID, tokenReq.NFTContractAddress)
 	}
 
-	// If we get here, all permissions are valid
-	return t.createAndReturnToken(c, pr, grantee)
+	return nil
 }
 
-func intArrayTo2BitArray(indices []int64, length int) (*big.Int, error) {
-	mask := big.NewInt(0)
+func evaluateCloudEvents(agreement map[string]map[string]*shared.StringSet, tokenReq *TokenRequest) error {
+	var err error
+	for _, req := range tokenReq.CloudEvents.Events {
+		if !common.IsHexAddress(req.Source) && req.Source != "*" {
+			err = errors.Join(err, fmt.Errorf("requested source %s invalid: must be %s or valid hex address", req.Source, tokenclaims.CloudEventTypeGlobal))
+			continue
+		}
 
-	for _, index := range indices {
-		if index < 0 || index >= int64(length) {
-			return big.NewInt(0), fmt.Errorf("invalid index %d. These must be non-negative and less than %d", index, length)
+		if len(req.IDs) == 0 {
+			err = errors.Join(err, fmt.Errorf("must request at least one cloudevent id or global access request (%s)", tokenclaims.CloudEventTypeGlobal))
 		}
-		mask.SetBit(mask, int(index*2), 1)
-		mask.SetBit(mask, int(index*2+1), 1)
-	}
 
-	return mask, nil
-}
+		grantedAggs, ok := agreement[req.EventType]
+		if !ok {
+			err = errors.Join(err, fmt.Errorf("lacking grant for requested event type: %s", req.EventType))
+			continue
+		}
 
-// validateAssetDID verifies that the provided DID matches the NFT contract address
-// and token ID specified in the permission token request.
-//
-// Parameters:
-//   - did: The decentralized identifier string to validate, typically in the format "did:nft:..."
-//   - req: The permission token request containing the NFT contract address and token ID to match against
-//
-// Returns:
-//   - bool: true if the DID is valid and matches the request parameters, false otherwise
-//   - error: An error describing why validation failed, or nil if validation succeeded
-func (t *TokenExchangeController) validateAssetDID(did string, req *TokenRequest) (bool, error) {
-	decodedDID, err := cloudevent.DecodeNFTDID(did)
-	if err != nil {
-		return false, fmt.Errorf("failed to decode DID: %w", err)
-	}
+		if globalAggs, grantedAll := grantedAggs[tokenclaims.CloudEventTypeGlobal]; grantedAll {
```

Comment:
nit: I always like to use ok for something existing in a map. But this is highly debatable so you can ignore
```suggestion
		if globalAggs, ok := grantedAggs[tokenclaims.CloudEventTypeGlobal]; ok {
```